server/src/main/java/com/flipkart/varadhi/web/v1/produce/ProduceHandlers.java

```diff
@@ -95,12 +102,21 @@ public void produce(RoutingContext ctx) {
         ProducerMetricsEmitter metricsEmitter = metricHandler.getEmitter(ctx.body().length(), produceAttributes);
 
         String varadhiTopicName = VaradhiTopic.buildTopicName(projectName, topicName);
+        // TODO(rl): handle for batch qps? Is it required
+        trafficAggregator.addTopicUsage(varadhiTopicName, ctx.body().length(), 1);
 
         // TODO:: Below is making extra copy, this needs to be avoided.
         // ctx.body().buffer().getByteBuf().array() -- method gives complete backing array w/o copy,
         // however only required bytes are needed. Need to figure out the correct mechanism here.
         byte[] payload = ctx.body().buffer().getBytes();
         Message messageToProduce = buildMessageToProduce(payload, ctx.request().headers(), produceIdentity);
+
+        Float suppressionFactor = suppressorHandler.getSuppressionFactor(varadhiTopicName);
+        if (!trafficAggregator.allowProduce(varadhiTopicName, suppressionFactor)) {
+            ctx.endRequestWithStatusAndErrorMsg(HTTP_RATE_LIMITED, "Rate limited");
+            return;
+        }
```

Comment:
not ideal usage. 
suppression factor looks implementation specific and should be visible here.

This class should only worry about adding the usage, and asking whether action is allowed.

Reply:
Made the changes. RL isAllowed API doesn't expose suppression factor.